### PR TITLE
wip: Use values file to supply values for prototypes

### LINKED ISSUE #550: Create values flag to supply values to `ks prototype use`
Values for prototypes can be complex. To allow for more flexibility, all user to use a value file to supply the options.

Usage: `ks prototype use <id> --values values-file`

Where `values-file` is file with keys and values:

```
key1=value1
key2=value2
```

### PR DESCRIPTION
**Pull request by bryanl**
```sh
ks generate deployed-service deployed-service5 --values-file /tmp/values-file
```

Fixes #550

Signed-off-by: bryanl bryanliles@gmail.com

### CODE REVIEW

File: pkg/clicmd/prototype_use.go [hunk 1]
```diff
@@ -75,6 +75,16 @@ expand prototypes into Jsonnet files.
 command line flags, such as ` + " `--image` " + `in the example above. Note that
 different prototypes support their own unique flags.
 
+You can also use a values file to supply values to the prototype. The format for the
+file is a key and value per line.
+
+		key1=value1
```
Comment (shomron) [inline pkg/clicmd/prototype_use.go]: We should be explicit about what characters are and aren't allowed in keys and values. For example, do we need to support `=` characters in values?

File: pkg/prototype/values.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+		return vff.Extract()
+	}
+
+	vff := newValuesFromFlags(proto, flags)
+	return vff.Extract()
+}
+
+type valuesFromFile struct {
+	fs    afero.Fs
```
Comment (shomron) [inline pkg/prototype/values.go]: Why is this needed? Is the `io.Reader` abstraction enough?

File: pkg/prototype/values_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+				"key2": `"default"`,
+			},
+		},
+		{
+			name:  "values: invalid",
+			p:     proto,
+			flags: BindFlags(proto),
+			args: []string{
+				"--name", "name",
+				"--values-file", "/files/invalid",
```
Comment (shomron) [inline pkg/prototype/values_test.go]: If `=` is prohibited in values, can we add a test case capturing that?

Comment (bryanl) [reply pkg/prototype/values_test.go]: It's not prohibited, but the string is split into two parts. I'll add a test case for this.

File: pkg/clicmd/prototype_use.go (hunk 1 above)
Comment (bryanl) [reply pkg/clicmd/prototype_use.go]: ksonnet accept any unicode character if the Param is designated as a string.

File: pkg/prototype/values_test.go (hunk 3 above)
Comment (shomron) [reply pkg/prototype/values_test.go]: It looks like we're explicitly checking for a two-way split, which would preclude additional `=`:

```
		parts := strings.SplitN(s, "=", 2)
		if len(parts) != 2 {
			return nil, errors.Errorf("%q is invalid. value format is key=value", s)
		}
```

Comment (bryanl) [reply pkg/prototype/values_test.go]: No, that's not how it would work. Here's an example:

https://play.golang.org/p/_uR5RufO667

Comment (shomron) [reply pkg/prototype/values_test.go]: Ah, so that is the difference between `Split` and `SplitN` 🙂
I misunderstood.

File: pkg/prototype/values.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+		quoted, err := param.Quote(val)
+		if err != nil {
+			return nil, err
+		}
+		values[param.Name] = quoted
+	}
+
+	if len(missingRequired) > 0 {
+		return nil, errors.Errorf("failed to instantiate prototype %q. The following required parameters are missing:\n%s", proto.Name, missingRequired.PrettyString(""))
```
Comment (shomron) [inline pkg/prototype/values.go]: `proto.Name` -> `vff.proto.Name`

File: pkg/prototype/values.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+import (
+	"bufio"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc/encoding/proto"
```
Comment (shomron) [inline pkg/prototype/values.go]: I believe this isn't needed.

File: Gopkg.lock [hunk 6]
```diff
@@ -773,6 +773,15 @@
   revision = "150dc57a1b433e64154302bdc40b6bb8aefa313a"
   version = "v1.0.0"
 
+[[projects]]
+  name = "google.golang.org/grpc"
```
Comment (shomron) [inline Gopkg.lock]: Was this brought in due to an over-eager `goimports`

File: pkg/prototype/values.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+		return nil, err
+	}
+
+	name, err := flags.GetString("name")
+	if err != nil {
+		return nil, errors.New("name for prototype is required")
+	}
+
+	if valuesFile != "" {
+		vff, err := newValuesFromFile(fs, p, name, valuesFile)
```
Comment (shomron) [inline pkg/prototype/values.go]: We should document somewhere that valuesFile will take precedence over bare flags.

File: pkg/prototype/prototype.go [hunk 8]
```diff
@@ -189,6 +200,25 @@ type ParamSchema struct {
...
+	var v string
+	if ps.Default != nil {
+		v = *ps.Default
+	}
+
+	if v == "" {
+		v = value
+	}
+
+	return ps.Quote(v)
```
[the remark this answers is not in the document]
Comment (bryanl) [reply pkg/prototype/prototype.go]: I think this should be filed as a bug and handled separately.

[the remark this answers is not in the document]
Comment (shomron) [reply pkg/prototype/prototype.go]: 👍

File: pkg/prototype/values.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+	values := map[string]string{}
+	missingRequired := ParamSchemas{}
+
+	for _, param := range vff.p.Params {
+		val, err := vff.flags.GetString(param.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		if param.Default != nil {
```
Comment (shomron) [inline pkg/prototype/values.go]: `param.IsRequired()`

File: pkg/prototype/values.go [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+		}
+
+		if parts[1] == "" {
+			return nil, errors.Errorf("%q is invalid. value format is key=value", s)
+		}
+
+		key := strings.TrimSpace(parts[0])
+		value := strings.TrimSpace(parts[1])
+
+		m[key] = value
```
Comment (shomron) [inline pkg/prototype/values.go]: Duplicate keys are not detected.

File: pkg/prototype/values.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+	if len(missingRequired) > 0 {
+		return nil, errors.Errorf("failed to instantiate prototype %q. The following required parameters are missing:\n%s", vff.p.Name, missingRequired.PrettyString(""))
+	}
+
+	return values, nil
+}
+
+func BindFlags(p *Prototype) *pflag.FlagSet {
+	fs := pflag.NewFlagSet("preview", pflag.ContinueOnError)
```
Comment (shomron) [inline pkg/prototype/values.go]: What does `preview` signify? Flags for the `prototype preview` command?

File: pkg/prototype/values_test.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+			},
+			{
+				Name:    "key2",
+				Type:    String,
+				Default: ksstrings.Ptr("default"),
+			},
+		},
+	}
+
+	cases := []struct {
```
Comment (shomron) [inline pkg/prototype/values_test.go]: We could use one more test to show precedence logic between bare flags and values file.

File: pkg/prototype/prototype.go [hunk 13]
```diff
@@ -189,6 +200,25 @@ type ParamSchema struct {
 	Type        ParamType `json:"type"`
 }
 
+// IsRequired returns if this parameter is required. Parameters are required
+// if the default value is nil.
+func (ps *ParamSchema) IsRequired() bool {
+	return ps.Default == nil
+}
+
+func (ps *ParamSchema) QuotedValue(value string) (string, error) {
```
Comment (shomron) [inline pkg/prototype/prototype.go]: I propose renaming this method `QuotedWithDefault` - as is the difference vs. `Quote` is not clear to the reader.

File: pkg/prototype/values.go [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+				missingRequired = append(missingRequired, param)
+				continue
+			}
+		}
+
+		if _, ok := values[param.Name]; ok {
+			return nil, errors.Errorf("prototype %q has multiple parameters with name %q", vff.p.Name, param.Name)
+		}
+
+		quoted, err := param.Quote(val)
```
Comment (shomron) [inline pkg/prototype/values.go]: Should probably call `QuotedValue` / `QuotedWithDefault`.

File: pkg/prototype/prototype.go (hunk 13 above)
Comment (shomron) [inline pkg/prototype/prototype.go]: Is it ever valid to set a parameter to an empty string? If so, we'll need a way to distinguish between that case and the case where no value was given and default should be used.

File: pkg/prototype/values.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+	}
+
+	return &valuesFromFile{
+		p:    p,
+		name: name,
+		r:    f,
+	}, nil
+}
+
+func (vff *valuesFromFile) Extract() (map[string]string, error) {
```
Comment (shomron) [inline pkg/prototype/values.go]: I think we should also error out on unexpected keys in the values file, where unexpected means no match in the schema for this prototype. The current state lets the default be used silently even though the user intended to override.

File: pkg/prototype/prototype.go (hunk 13 above)
Comment (shomron) [reply pkg/prototype/prototype.go]: related: https://github.com/ksonnet/ksonnet/issues/273
